extmod/modopenamp.c

```diff
@@ -0,0 +1,399 @@
+    rpmsg_deinit_vdev(&virtio_device->rvdev);
+    metal_finish();
+    MP_STATE_PORT(virtio_device) = NULL;
+    return mp_const_none;
+}
+static MP_DEFINE_CONST_FUN_OBJ_1(virtio_dev_deinit_obj, virtio_dev_deinit);
+
+static const mp_rom_map_elem_t virtio_dev_locals_dict_table[] = {
+    { MP_ROM_QSTR(MP_QSTR___name__), MP_ROM_QSTR(MP_QSTR_VirtIODev) },
+    { MP_ROM_QSTR(MP_QSTR___del__), MP_ROM_PTR(&virtio_dev_deinit_obj) },
+};
+static MP_DEFINE_CONST_DICT(virtio_dev_locals_dict, virtio_dev_locals_dict_table);
+
+MP_DEFINE_CONST_OBJ_TYPE(
+    virtio_dev_type,
+    MP_QSTR_VirtIODev,
+    MP_TYPE_FLAG_NONE,
+    locals_dict, &virtio_dev_locals_dict
+    );
+
+// ###################### RPMsg Endpoint class ######################
+typedef struct _endpoint_obj_t {
+    mp_obj_base_t base;
+    mp_obj_t name;
+    mp_obj_t callback;
+    struct rpmsg_endpoint ep;
+} endpoint_obj_t;
+
+const mp_obj_type_t endpoint_type;
+
+static int endpoint_recv_callback(struct rpmsg_endpoint *ept, void *data, size_t len, uint32_t src, void *priv) {
+    debug_printf("endpoint_recv_callback() message received src: %lu msg len: %d\n", src, len);
+    endpoint_obj_t *self = metal_container_of(ept, endpoint_obj_t, ep);
+    if (self->callback != mp_const_none) {
+        mp_call_function_2(self->callback, mp_obj_new_int(src), mp_obj_new_bytearray_by_ref(len, data));
+    }
+    return 0;
+}
+
+static mp_obj_t endpoint_send(uint n_args, const mp_obj_t *pos_args, mp_map_t *kw_args) {
+    enum { ARG_src, ARG_dest, ARG_timeout };
+    static const mp_arg_t allowed_args[] = {
+        { MP_QSTR_src, MP_ARG_INT | MP_ARG_KW_ONLY,  {.u_int = -1 } },
+        { MP_QSTR_dest, MP_ARG_INT | MP_ARG_KW_ONLY,  {.u_int = -1 } },
+        { MP_QSTR_timeout, MP_ARG_INT | MP_ARG_KW_ONLY,  {.u_int = 0 } },
+    };
+
+    // Parse args.
+    mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
+    mp_arg_parse_all(n_args - 2, pos_args + 2, kw_args, MP_ARRAY_SIZE(allowed_args), allowed_args, args);
+
+    endpoint_obj_t *self = MP_OBJ_TO_PTR(pos_args[0]);
+
+    if (is_rpmsg_ept_ready(&self->ep) == false) {
+        mp_raise_msg(&mp_type_OSError, MP_ERROR_TEXT("Endpoint not ready"));
+    }
+
+    uint32_t src = self->ep.addr;
+    if (args[ARG_src].u_int != -1) {
+        src = args[ARG_src].u_int;
+    }
+
+    uint32_t dest = self->ep.dest_addr;
+    if (args[ARG_dest].u_int != -1) {
+        dest = args[ARG_dest].u_int;
+    }
+
+    mp_buffer_info_t rbuf;
+    mp_get_buffer_raise(pos_args[1], &rbuf, MP_BUFFER_READ);
+    debug_printf("endpoint_send() msg len: %d\n", rbuf.len);
+
+    int bytes = 0;
+    uint32_t timeout = args[ARG_timeout].u_int;
+    for (mp_uint_t start = mp_hal_ticks_ms(); ; mp_hal_delay_ms(10)) {
+        bytes = rpmsg_send_offchannel_raw(&self->ep, src, dest, rbuf.buf, rbuf.len, false);
+        if (bytes > 0) {
+            break;
+        }
+        if (timeout > 0 && (mp_hal_ticks_ms() - start > timeout)) {
```

Comment:
I suggest `timeout` is signed `mp_int_t` and the first condition here is `timeout >= 0`, so that a negative number means wait forever.

Reply:
Timeout of 0 or -1 will be both be blocking, do you mean break on `timeout == 0` like so:
```C
    for (mp_uint_t start = mp_hal_ticks_ms(); ;) {
        bytes = rpmsg_send_offchannel_raw(&self->ep, src, dest, rbuf.buf, rbuf.len, false);
        if (bytes > 0) {
            break;
        }
        if (timeout == 0) {
            break;
        }
        if (timeout > 0 && (mp_hal_ticks_ms() - start > timeout)) {
            mp_raise_msg(&mp_type_OSError, MP_ERROR_TEXT("timeout waiting for a free buffer"));
        }
        MICROPY_EVENT_POLL_HOOK
    }
```
I think I should also return the number of bytes sent...

Reply:
One minor change to fix a corner case, events need to be polled at least once to make sure the rproc notify node is scheduled, if send is called in a tight loop with no sleep.